audit/audit.go

```diff
@@ -0,0 +1,272 @@
+}
+
+// Client is an audit server smart client. This handles any connection problems silently, while providing ways to
+// detect problems through whatever alerting method is desired.
+type Client struct {
+	// client holds the current audit client. If this is nil, it means we are trying to get a new client.
+	// When that occurs, we will do a noop on Send().
+	client *base.Client
+	// replaceClient is used to ensure we only replace the client once. This is needed because we can have multiple
+	// goroutines trying to replace the client at the same time.
+	replacingClient atomic.Bool
+
+	// create is the function that creates a new connection to the remote audit server. This is used when the
+	// connection is broken.
+	create CreateConn
+	// backoff is the exponential backoff for making a new connection to the remote audit server.
+	backoff *exponential.Backoff
+	policy  exponential.Policy
+
+	// notifier is the channel that will receive errors when the connection to the remote audit server is broken or
+	// the queue is full. If this channel is full, the error will be dropped.
+	notifier chan NotifyError
+	// auditOptions are the options for the underlying base.Client.
+	auditOptions []base.Option
+
+	// replaceBackoffRunner is a function that replaces the audit client with a new one using an exponential backoff.
+	// In production, this uses .replaceBackoff. In testing this can be replaced with a function that does nothing.
+	replaceBackoffRunner func(ctx context.Context) error
+	// sendRunner is a function that sends a message to the remote audit server.
+	// In production, this uses .client.Sent(). In testing this can be replaced with a function that does nothing.
+	sendRunner func(ctx context.Context, msg msgs.Msg) error
+}
+
+// CreateConn is a function that creates a connection to a remote audit server.
+type CreateConn func() (conn.Audit, error)
+
+// Option is an option for the smart client.
+type Option func(*Client) error
+
+// WithAuditOptions sets the audit options for the underlying audit client.
+func WithAuditOptions(options ...base.Option) Option {
+	return func(c *Client) error {
+		c.auditOptions = options
+		return nil
+	}
+}
+
+// WithBackoffPolicy sets the exponential backoff policy for making a new connection to the remote audit server.
+// By default this uses the default backoff settings from the exponential package.
+func WithExponentialBackoff(policy exponential.Policy) Option {
+	return func(c *Client) error {
+		c.policy = policy
+		return nil
+	}
+}
+
+// New creates a new smart client to the remote audit server. You can get a CreateConn
+// from conn.NewDomainSocket(), conn.NewTCP() or conn.NewNoop(). The last one is useful for testing
+// when you don't want to send logs anywhere.
+func New(create CreateConn, options ...Option) (*Client, error) {
+	connect, err := create()
+	if err != nil {
+		return nil, err
+	}
+
+	if runtime.GOOS != "linux" && !testing.Testing() {
+		if connect.Type() != conn.TypeNoOP {
```

Comment:
minor, style: Why not 3 `&&` rather than 2 `&&` and a nesting?

feel free to ignore.

Reply:
For me these things are always about how readable I find it.

I don't find: `runtime.GOOS != "linux" && !testing.Testing() && connect.Type() != conn.TypeNoOp` easily readable.
I could do 3 tiers, but I think that is also less readable.  I felt that I just had enough in the top line that was readable.

In reality, I do this using the force more than some rule.  And it applies to how I read code more than how you might.

Reply:
Btw, I will do it any way you prefer.

Reply:
What you have is fine by me!